src/executor.c

```diff
@@ -0,0 +1,233 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv-common.h"
+
+#if !defined(_WIN32)
+# include "unix/internal.h"
+#endif
+
+#include <stdlib.h>  /* abort() */
+
+#include <stdio.h> /* Debug */
+
+static uv_executor_t *executor = NULL;
+
+static uv_once_t once = UV_ONCE_INIT;
+static volatile int initialized = 1; /* Protected by once in
+                    uv_executor_queue_work, but not in uv_replace_executor. */
+
+uv_executor_t * uv__executor(void) {
+  return executor;
+}
+
+/* CB: Executor has finished this request. */
+static void uv__executor_done(uv_work_t* req) {
+  uv_mutex_lock(&req->loop->wq_mutex);
+
+  /* Place in the associated loop's queue.
+   * NB We are re-purposing req->work_req.wq here.
+   * This field is also used by the default executor, but
+   * not after the default executor has called done() to get here. */
+  QUEUE_INSERT_TAIL(&req->loop->wq, &req->work_req.wq);
+
+  /* Signal to loop that there's a pending task. */
+  uv_async_send(&req->loop->wq_async);
+
+  uv_mutex_unlock(&req->loop->wq_mutex);
+}
+
+/* CB: Event loop is handling completed requests. */
+void uv__executor_work_done(uv_async_t* handle) {
+  struct uv__work* w;
+  uv_work_t* req;
+  uv_loop_t* loop;
+  QUEUE* q;
+  QUEUE wq;
+  int err;
+
+  /* Grab this loop's completed work. */
+  loop = container_of(handle, uv_loop_t, wq_async);
+  uv_mutex_lock(&loop->wq_mutex);
+  QUEUE_MOVE(&loop->wq, &wq);
+  uv_mutex_unlock(&loop->wq_mutex);
+
+  /* Handle each uv__work on wq. */
+  while (!QUEUE_EMPTY(&wq)) {
+    q = QUEUE_HEAD(&wq);
+    QUEUE_REMOVE(q);
+
+    w = container_of(q, struct uv__work, wq);
+    req = container_of(w, uv_work_t, work_req);
+    err = (req->work_cb == uv__executor_work_cancelled) ? UV_ECANCELED : 0;
+
+    uv__req_unregister(req->loop, req);
+
+    if (req->after_work_cb != NULL)
+      req->after_work_cb(req, err);
+  }
+}
+
+int uv_replace_executor(uv_executor_t* _executor) {
+  /* Reject if no longer safe to replace. */
+  if (!initialized)
+    return 1;
+
+  /* Check validity of _executor. */
+  if (_executor == NULL)
+    return 1;
+  if (_executor->submit == NULL)
+    return 1;
+
+  /* Set private fields. */
+  _executor->done = uv__executor_done;
+
+  /* Replace our executor. */
+  executor = _executor;
+
+  return 0;
+}
+
+static void uv__executor_init(void) {
+  /* Assign executor to default if none was set. */
+  if (executor == NULL) {
+    assert(!uv_replace_executor(uv__default_executor()));
```

Comment:
> Don't perform actions with side effects inside an assert().

OK

> I can't find where uv__default_executor() is defined

It's defined in `src/threadpool.c` since that is where the default executor is implemented. It basically returns a pointer to the `uv_executor_t default_executor` owned by `src/threadpool.c`.